Failsafe - New failsafe functionality (Phase 1)

ProDrone opened the pull request:
__This pull request is to replace and add all the features (and some more) of #185__

### Changes in functionality
* Added a Failsafe flightmode.

* Added AUX switch (RC controlbox) to manually activate failsafe mode.

* Prevent arming when failsafe mode via AUX switch is active (safety issue).

* Added a CLI variable `failsafe_kill_switch` for instant disarm (__KillswitchEvent__).
When set to 1 (0 is default), the failsafe switch will instantly disarm
instead of executing the auto-land procedure. Arming is NOT locked after
this, so the craft could be re-armed if needed.
This is intended for racing quads where damage and danger must be
minimized in case of a pilot error.

* Make failsafe disarm if motors armed and throttle was LOW (2D & 3D) for `failsafe_throttle_low_delay` time (__JustDisarmEvent__).
 Applied code changes to effectively add PR #717.

* Updated the failsafe documentation.

* Re-enable arming at end of failsafe procedure (alternative implementation for #1181).
At the end of a handled failsafe event, that means: auto-landing, __JustDisarmEvent__ or __KillswitchEvent__, the RX link is monitored for valid data. Monitoring is a part of the failsafe handling, which means the craft is still in failsafe mode while this is done. Arming is re-enabled (allowed) when there is a valid RX link for more then XX seconds, where XX depends on the handled event like this:
```p
XX =  1  second after KillswitchEvent
XX =  3 seconds after JustDisarmEvent
XX = 30 seconds after auto landing
```
__NOTE:__
When armed via an AUX switch, you will have to switch to the disarmed position at the very end to be able to re-arm. _The failsafe mode will not end until you do._

ethomas997 commented:
Could you post a summary of the changes in functionality for this PR?

--ET

ProDrone commented:
@ethomas997 It is in the description of the last 5 commits above. Basically all you asked for and a bit more...

borisbstyle commented:
This feature sounds like dangerous when using pid_at_min_throttle, which is enabled by default
---->>>>>
Make failsafe disarm if motors armed and throttle low for 3 seconds. Applied code changes to effectively add PR #717

ProDrone commented:
@hydra It is now rebased on a recent master. I modified the testcode of __flight_failsafe_unittest.cc__
so that it passes Travis CI build. It is just running the existing tests with no test added for the new features. I need to learn a bit more in order to make those and hope you will take a look at it.

Also modified testcode __rc_controls_unittest.cc__ since it was not compiling (the initialiser for the new rcYawExpo8 member in controlRateConfig was missing).

hydra commented:
@ProDrone I'll look at this soon, it's high on my priority list.

I have done quite a bit of flying with the current master code and it looks like many others have too with no reported issues on the existing re-worked failsafe code.  I wanted master to be stable before merging any more changes so as not to build on something that might be broken.   I think I am satisfied the current master is in a state where this can be added.

hydra commented:
@ProDrone I might get to look at this this weekend and merge it for 1.9.0.  If not then 1.9.0 will go out without it but this will be the first thing that gets my attention for 1.10.0.  Binaries for SPRacingF3 need to go to manufacturing :smile:

ProDrone commented:
Bit of topic maybe, but i hope the SPRacingF3 will become a big success for you. I saw some critical notes at http://www.rcgroups.com/forums/showpost.php?p=31607200&postcount=53 and also about the price compared to other boards. People seem to forget that the Naze32 is also at rev 5. Evolution i mean. I think if changes are demanded you will adapt, right?

borisbstyle commented:
That guy is complaining about connectors lol.
Isn't openpilot running the same connectors and nobody ever made a point because of that :).

ProDrone commented:
It is okay if __some__ complain i guess you can never satisfy them all. And shall we remove those comments from here? They are not on topic or useful to others.

ProDrone commented:
@hydra Starting with #781 the failsafe settings are moved from profile to masterconfig.
What do you think about the validity of the point made in this post?:
http://www.rcgroups.com/forums/showpost.php?p=31619301&postcount=7655

_"These settings are profile dependent which is great, my profile 0 is for use with 5040 props, while profile 2 is for use with 5045 bullnose. They both have different failsafe_throttle values, otherwise by using more powerful props I could end up with a throttle too high and a fly-away on failsafe."_

ethomas997 commented:
I did some bench testing on a copter with an RX_SERIAL satellite receiver.  Overall, worked as advertised.  One thing I noticed -- regarding the "Make failsafe disarm if motors armed and throttle low for 3 seconds" functionality -- the buffer time is much less than 3 seconds.  This is probably because those functions are not called at 50Hz.

I think that if the failsafe auto-landing was triggered via the aux-mode switch then PREVENT_ARMING should not be set.

Seems like the 'failsafe_kill_switch' function is doing the same thing as the aux-mode-arming switch when 'disarm_kill_switch' is enabled.  Is the idea to have a kill switch when stick arm/disarm is used?  If so, we could look at adding a new option to support using both the aux-mode-arming switch and stick arm/disarm together.

--ET

ProDrone commented:
@ethomas997 Thanks for testing. Timing is still an unsolved issue (like in the current implementation). My intention is to solve this once it is merged with master. Although times runs faster then advertised, you can still set another value to get the desired timing.

When failsafe auto-landing is completed via aux, it was probably because of some problematic situation. Your copter may be somewhere out of sight but on the ground. That is why i like PREVENT_ARMING set.

Yes, `failsafe_kill_switch` is intended in combination with stick arming (`disarm_kill_switch` is related but something different). How should your proposed new option work?

ethomas997 commented:
@ProDrone  Regarding PREVENT_ARMING, I was only thinking of aux-mode-switch failsafe as being used for testing the functionality.

For a new option to support using both the aux-mode-arming switch and stick arm/disarm together, what I'm thinking is that the switch would then function like an "arming enable," where the copter can only be stick-armed if the switch is 'on', and moving the switch to 'off' would immediately disarm the copter.  (Stick-disarm would still work the same.)  A new CLI variable would enable this (call it "stick_switch_arming"?) and the aux-mode-arming switch would be configured like it is currently.

The way I'd see it being used is that the pilot would generally have the aux-mode-arming switch always in the 'on' position.  Normal arming and disarming would be done via the stick.  However, when there's trouble during a flight, hitting the switch would quickly kill the motors.

I can do a PR for this...

--ET

ProDrone commented:
@ethomas997 It is useful for testing, but also intended as a kind of panic switch. You have the option to try to auto-land (normal operation) or instant kill motors and disarm (`failsafe_kill_switch` set to 1).
The option as you describe is partly similar (the kill and disarm part). and on top of that it disables the stick-arming, if i see it right. What is the added benefit? (sorry i don't see it yet).

ethomas997 commented:
I see the "stick_switch_arming" concept as fitting in better with the current functionality.  A new "failsafe_kill_switch" setting would duplicate a lot of what "disarm_kill_switch" already does.

When the aux-mode-switch failsafe auto-landing is active it disables all the controls, which could lead to other kinds of troubles.  Seems like using an aux-mode switch to activate 'angle' mode would be a better panic switch for those situations.

I guess another option would be to implement both "failsafe_kill_switch" and "stick_switch_arming".

--ET

ProDrone commented:
@ethomas997 Another reason why i added an aux switch to trigger failsafe was that some receivers cannot be configured to stop sending valid output when the radio connection is lost. They can only be programmed to let channels go to a predefined state. Later i added the `failsafe_kill_switch`, just because it can be handled in the failsafe module and i like the idea of an instant kill feature.

I also like the idea you propose here, but it will be assigned to a different switch and is not part of the failsafe handling, but the arming/disarming procedure. Maybe you should make a PR for it.

hydra commented:
@ProDrone You you rebase this PR on latest master please?

ProDrone commented:
@hydra Just rebased, now testing...

ProDrone commented:
@hydra Rebased on 1c1bdfaa1fe8d1dd84c0441e1a592976bc77d0e7 requires testing after #995 has been merged (but i'm working on that one right now).

joshuabardwell commented:
>Another reason why i added an aux switch to trigger failsafe was that some receivers cannot be configured to stop sending valid output when the radio connection is lost. 

This may be too obvious to even mention, but my first thought was that it would allow testing a failsafe situation where the receiver is disconnected or powered down without having to physically unplug your receiver, which can be difficult, require disassembling the copter, and simply can't be done in flight.

ProDrone commented:
Some receivers can only be programmed to let channels go to a specified value when a failsafe condition occurs. That value is still in the valid range and as such cannot be classified as invalid by the flight-controller. When you assign the value for the on position, to the channel that is controlled by the failsafe-switch, then you can still control what happens when a failsafe condition occurs. In addition you can use this switch to test the failsafe handling.

ProDrone commented:
@hydra
* Rebased.
* Testing depends on code from #995.
* CI build failed because of modified behavior like this: when throttle has been below min_throttle for `failsafe_throttle_low_delay` period and failsafe condition occurs, the craft is just disarmed to prevent spin up on the ground.
* I don't know how to properly adapt the unit test, so these fail:

`[  FAILED  ] FlightFailsafeTest.TestFailsafeDetectsRxLossAndStartsLanding`
`[  FAILED  ] FlightFailsafeTest.TestFailsafeCausesLanding`

ProDrone commented:
@hydra Finally picking up this PR for testing again...
This depended for a long time on code from #995 -> #1190 -> #1227 -> #1203 -> etc...
When ready for merging i will let you know.

* Rebased.
* CI build failed because of modified behavior like this: when throttle has been below min_throttle for `failsafe_throttle_low_delay` period and failsafe condition occurs, then the craft is just disarmed to prevent spin up on the ground.
* I don't know how to properly adapt the unit test, so these fail:

`[  FAILED  ] FlightFailsafeTest.TestFailsafeDetectsRxLossAndStartsLanding`
`[  FAILED  ] FlightFailsafeTest.TestFailsafeCausesLanding`

ProDrone commented:
@hydra 

* Rebased on c49bd407bfe716b2566ec119953941a3c2283e8d

* Added feature: Re-enable arming at end of failsafe procedure (alternative implementation for #1181).
See commit- or first on top message.

Now even more test failed :-( I need some help with that please...

ctzsnooze commented:
Thanks ProDrone for your determined work!  This functionality is really important.

As you know I've been wondering about the configuration of the failsafe landing phase itself.  At the moment the user only gets level mode at a set throttle for a set time.  It might be preferable for some users to control channel values more specifically for the landing phase, e.g. to activate GPS return to home in the landing phase, or to activate a channel that makes the buzzer loud, or activate a special flashing LED or something weird.  The current rxfail solution allows this level of flexibility during the failsafe detection phase, but we currently do not have the same level of control in the landing phase.

So I was wondering if we could add an additional column of values in rxfail, which, if not zero, set the value to be sent *during the failsafe landing phase* on each channel (or aux channel).  That would provide full flexibility for the user during both failsafe detection and failsafe landing phases.  They could for example hold values during detection, then activate GPS return to home with a loud buzzer.  

Depending on programming challenge, it may be necessary to have a CLI entry for the user to get default classical failsafe (level mode only) or alternatively their personal rxfail-configured values.

It is true that the advanced user who wants something other than level mode in failsafe landing could configure their Rx channels individually to send GPS RTH and disable the CF failsafe, so this suggestion is not essential.

Anyhow I am wondering about your thoughts on this idea.  I don't mind if it is ignored.  It just seems to me to provide a more complete failsafe solution.

Secondly, I wonder if the delay time for re-arming after failsafe landing is complete could be say 15s not 30s?

ProDrone commented:
I see it as the next logical step, _after this proposed functionality is accepted and stable_, to focus on PH and RTH.

`Rxfail` is intended to cover the loss of the RC link even with `feature FAILSAFE` disabled. So a separate failsafe configuration must be created to handle this.

The 30 seconds for re-arming is a bit long because on serial RX's _time_ runs faster (i had to set something in the middle). This is a problem that needs to be solved next. Btw. 30 seconds are over before you know it, when climbing a tree or a fence to recover your quad ;)

ctzsnooze commented:
Thanks for the clarifications.  And yes, most of my lost stuff has been *at least* 30s away!  :-)

ProDrone commented:
@hydra

This depended for a long time on code from #995 -> #1190 -> #1227 -> #1203 -> etc...
See your comment of 11 and 14 May above...

I hope it will make it this time in release 1.10.0 (before the world changes again).

* Squashed and rebased on c49bd407bfe716b2566ec119953941a3c2283e8d

* Added feature: Re-enable arming at end of failsafe procedure (alternative implementation for #1181).
See commit- or first (top) message.

* Bench tested all cases of these features up to and including this last one.

* Adapted the unit test too! :smile: 

* __This code is ready for merging__ and provides quite a lot of improvements.

===

There are some compile time warnings in display.c:

`%% display.c
./src/main/io/display.c: In function 'updateFailsafeStatus':
./src/main/io/display.c:221:5: warning: enumeration value 'FAILSAFE_RX_LOSS_MONITORING' not handled in switch [-Wswitch]
     switch (failsafePhase()) {
     ^
./src/main/io/display.c:221:5: warning: enumeration value 'FAILSAFE_RX_LOSS_RECOVERED' not handled in switch [-Wswitch]`

ProDrone commented:
@hydra Please consider issue #1254, solved with PR #1255 together with this PR.

ProDrone commented:
@hydra Please consider issue #1254, solved with PR #1255 _and_ issue #1257, solved with PR #1258 together with this PR.

hydra commented on src/main/flight/failsafe.c:
```diff
@@ -159,50 +171,103 @@ void failsafeUpdateState(void)
-
                 if (failsafeShouldHaveCausedLandingByNow() || !armed) {
-
+                    failsafeState.receivingRxDataCounts = 1500; // require 30 seconds of valid rxData at 50Hz
```
magic number alert!  use a constant and if possible build it using other constants.

`#define MY_CONSTANT (30 * RX_FREQUENCY)`

using a counter is introducing another bug that needs to be fixed because of the update rate of rx is not always 50hz.  See other FIXMEs.  Use a fixed time `shouldDoSomethingAt = ...` or record a start time and use an interval that can be extended.  `somethingHappenedAt = ...;   interval = 30_SECONDS`  ...  `if (something) interval += 5_SECONDS`.

or something.

hydra commented on src/main/flight/failsafe.c:
```diff
@@ -159,50 +171,103 @@ void failsafeUpdateState(void)
             case FAILSAFE_LANDED:
+                ENABLE_ARMING_FLAG(PREVENT_ARMING); // To prevent accidently rearming by an intermittent rx link
+                mwDisarm();
+                failsafeState.receivingRxDataCounter = failsafeState.receivingRxDataCounts; // set required counts of valid rxData
```
See note about avoiding counter use above

hydra commented on src/main/flight/failsafe.c:
```diff
@@ -220,5 +285,6 @@ void failsafeUpdateState(void)
 void failsafeOnRxCycleStarted(void)
 {
+    // FIXME Timing is based on the assumption that this function is called at 50Hz
     failsafeState.counter++;
```
I think this PR should address this FIXME.

hydra replied:
We *could* address the FIXME as the next failsafe PR, but I'd rather the FIXME was fixed before merging this as it will be cleaner that way around.  Generally speaking bugs should always be fixed before adding new code.  Building on code that has bugs is not good practice, leads to unpredictable behavior and costs more time in the long run.

hydra commented on src/test/unit/flight_failsafe_unittest.cc:
```diff
@@ -226,6 +236,7 @@ TEST(FlightFailsafeTest, TestFailsafeCausesLanding)
 {
     // given
     int callsToMakeToSimulateFiveSeconds = FAILSAFE_UPDATE_HZ * 5;
+    int callsToMakeToSimulateFiftySeconds = FAILSAFE_UPDATE_HZ * 50;
```
Again we can see that the FIXME and the use of the counter is having a negative impact even on the tests, the test should be manipulating the response the code under test gets from a call to `micros()` or `millis()`

hydra commented on src/test/unit/flight_failsafe_unittest.cc:
```diff
@@ -256,16 +267,146 @@ TEST(FlightFailsafeTest, TestFailsafeCausesLanding)
+    EXPECT_EQ(FAILSAFE_IDLE, failsafePhase());
+    EXPECT_EQ(1, CALL_COUNTER(COUNTER_MW_DISARM)); // disarm not called repeatedly.
+    EXPECT_FALSE(ARMING_FLAG(PREVENT_ARMING));
 }
```
counters everywhere!  yes.  the failsafe needs to move away from counters and release itself from expected call frequency problem it has.

hydra commented on src/test/unit/rc_controls_unittest.cc:
```diff
@@ -445,6 +465,26 @@ static const adjustmentConfig_t rateProfileAdjustmentConfig = {
+    adjustmentStateMask = 0;
+    memset(&adjustmentStates, 0, sizeof(adjustmentStates));
+
     int adjustmentIndex = 3;
```
duplicate test setup code can be removed.  See `RcControlsAdjustmentsTest::SetUp()`  It looks like there might be some duplicate configuration of controlRateConfig anyway.

hydra commented on src/test/unit/rx_ranges_unittest.cc:
```diff
@@ -39,6 +41,8 @@ extern uint16_t applyRxChannelRangeConfiguraton(int sample, rxChannelRangeConfig
 
 TEST(RxChannelRangeTest, TestRxChannelRanges)
 {
+    rcModeActivationMask = 0;                   // de-activate BOXFAILSAFE
```
Perhaps a constant instead of magic number `0` and a comment?

hydra commented on src/test/unit/rx_rx_unittest.cc:
```diff
@@ -46,6 +48,7 @@ TEST(RxTest, TestValidFlightChannels)
 {
     // given
     memset(&testData, 0, sizeof(testData));
+    rcModeActivationMask = 0;                   // de-activate BOXFAILSAFE
```
Same.  See https://github.com/cleanflight/cleanflight/pull/852/files#r38135614

ProDrone commented:
@hydra 

- Reworked the code from counter-based to time-based (no more counters).
- Bench tested.
- Adapted failsafe related unittests from counter-based to time-based.
- Squashed and rebased on 1a15e5aa0815c5c672ea66fa8369a97a88c85e7b
- __Ready for merge...__

===

__Please also consider these:__
- issue #1254, solved with PR #1255;
- issue #1257, solved with PR #1258;
- issue  #1262, solved with PR #1263

hydra commented:
great, reviewing now.

hydra commented:
Ok, I'm going to merge this.

There are significant improvements to functionality, bug fixes and cleanups.  Nice work!
